### DISCUSSION TITLE
Several questions about the tuner

### DISCUSSION THREAD

**Discussion by vocna** [thetwom/Tuner#69] [Q&A]
What _Tolerance_ setting exactly does? I mean in technically, how it works?

What exactly is pitch result? E.g. _Moving average_ and _Jump to new pitch_ talks about them, but what is the defition of pitch result?

What exactly is background noise? I mean what is the defition of background noise for the app?

How _spectrum_ and _autocorrelation_ works? And what is the difference between them?
What is _Windowing function_?

For what _Capture in scientific mode_ is used?

**Comment by thetwom** [comment]
Hi, thanks for the questions:

**Tolerance**
This defines the two lines above and below your target note in the pitch history window. So if you tune between these two lines, you are maximum 5 (or whatever value you chose) cent of. Actually 5cent is quite accurate. 

**Pitch**
The pitch result refers to the most probable note frequency, which is currently detected. We cannot simply say, the "current frequency", since when an instrument plays a note, it almost never plays pure frequencies (as e.g. 440Hz for A4). But mostly we also find "higher harmonics", e.g. 880Hz, 1220Hz, ... for A4. In the end you have to decide on something like the most basic frequency. And this is, what is meant with pitch. 

**Background noise**
If you don't currently play a note, the app will still detect noise. It could still try to find notes from this, but it makes not much sense. The question is how to find out if currently the instrument is played or not. At the moment this is done by checking if we have very high noise level compared to the overall recorded sound. Noise is a sound which consists of all frequencies, not just specific ones. In the app, background noise is the first value of the autocorrelation divided by the total amount of "sound energy", but this get quite technical ...

**Spectrum**
We record sound in the so-called "time domain". Meaning that you e.g. record 0.1s and then get many "pressure disturbance" values (this is the sound) for different times during these 0.1s. This can now transformed to a so called frequency spectrum to see which frequencies the signal contains (Fourier transformation). Inside this spectrum you can find the frequencies with high values and based on that, get the "pitch".

**Autocorrelation**
Autocorrelation is if you take the time signal from above, then shift it by a very small value and compare how similar the original and the shifted signals are. Then you shift further and compare again, and so on. For wavy signals, at certain shifts the signal compares better to the original signal. This tells then something about the underlying pitch. This is a more robust method for detecting pitches than the spectrum. The spectrum is still helpful, for improving the accuracy of the detected pitch, but this is another story ....

**Windowing**
This gets very technical. But if we take again the 0.1s signal, we can scale the start and end of the signal down before computing a spectrum. This can help to avoid getting incorrect frequencies. At the moment I am uncertain if this actually is important at all for a tuner. The "window" defines the shape how the ends are scaled down. 

**Capture** 
If you are in scientific mode and have this setting on, then in main window you will see a capture button at the bottom. This allows you to save the recorded signal of the last 1,2,3,... seconds in a wav-file. I introduced it to allow users to easily record a signal and then provide the sound file for analysis (e.g. the tuner is not working well for them). But up to now, this was only used by me ...

Please let me know, if this helps ...

**Reply by vocna** [reply]
> **Tolerance**
> This defines the two lines above and below your target note in the pitch history window. So if you tune between these two lines, you are maximum 5 (or whatever value you chose) cent of. Actually 5cent is quite accurate.

I know that, but is tolerance 1 1Hz, 5Hz, half tone or what? How much is tolerance 1?

> **Pitch**
> The pitch result refers to the most probable note frequency, which is currently detected. We cannot simply say, the "current frequency", since when an instrument plays a note, it almost never plays pure frequencies (as e.g. 440Hz for A4). But mostly we also find "higher harmonics", e.g. 880Hz, 1220Hz, ... for A4. In the end you have to decide on something like the most basic frequency. And this is, what is meant with pitch.

Yeah, I know, but question was about pitch results. What is pitch result? I mean e.g. text "After 5 similar pitch results". How long one sample is? Is it, what is strongest Hz in 0,05s time? So if for 0,05s strongest Hz is 440, then next 0,05s, again, again and again, the monitor will jump to a¹?

Side question about your answer:
I know, yhe defition of a pitch is strongest Hz in tone, so what makes so hard to recognize it? Just look recorded sample and pick the strongest Hz in it?

> **Background noise**
> If you don't currently play a note, the app will still detect noise. It could still try to find notes from this, but it makes not much sense. The question is how to find out if currently the instrument is played or not. At the moment this is done by checking if we have very high noise level compared to the overall recorded sound. Noise is a sound which consists of all frequencies, not just specific ones. In the app, background noise is the first value of the autocorrelation divided by the total amount of "sound energy", but this get quite technical ...

Why? I mean, in tone is always one strongest Hz and we can always recognize it, right?
So, is _Maximum allowed background noise_ the sensitivity of the tuner search for pitch from tone it receives? If I set the setting 0% the tuner will always tell the current pitch (strogest Hz) and if 100% it will never tell any the current pitch?

I'd like to hear technical explanation :)

> **Spectrum**
> We record sound in the so-called "time domain". Meaning that you e.g. record 0.1s and then get many "pressure disturbance" values (this is the sound) for different times during these 0.1s. This can now transformed to a so called frequency spectrum to see which frequencies the signal contains (Fourier transformation). Inside this spectrum you can find the frequencies with high values and based on that, get the "pitch".

What setting defines used time domain in that monitor?
Okay. So, the bottom (0Hz...1000Hz...) of the monitor tells current frequencies/Hz:s, vertical tells strength of them Hz, right? But there can I see timeline of the time domain? Or does the monitor shows real-time current tone/sound (and not time domain) and strengths of Hz:s in it? Is the current pitch always Hz that is highest in vertical in this monitor?
Does _Maximum allowed background noise_ or other settings affect to this monitor?

> **Autocorrelation**
> Autocorrelation is if you take the time signal from above, then shift it by a very small value and compare how similar the original and the shifted signals are. Then you shift further and compare again, and so on. For wavy signals, at certain shifts the signal compares better to the original signal. This tells then something about the underlying pitch. This is a more robust method for detecting pitches than the spectrum. The spectrum is still helpful, for improving the accuracy of the detected pitch, but this is another story ....

Now I can't understand... The bottom of this monitor is also 0Hz...1000Hz... like in spectrum monitor so I assume that them tells Hz of showed frequency above, vertical still tells strength of Hz:s, right?
Why this monitor manytimes looks like a snake? Spectrum monitor shows different strength values, why?
I assume that vertical line tells the most strong Hz also in that monitor, but what horizontal line tells?

> **Windowing**
> This gets very technical. But if we take again the 0.1s signal, we can scale the start and end of the signal down before computing a spectrum. This can help to avoid getting incorrect frequencies. At the moment I am uncertain if this actually is important at all for a tuner. The "window" defines the shape how the ends are scaled down.

I can't see the difference between _No windowing function_, _Hamming window_ and _Hann window_.

> **Capture**
> If you are in scientific mode and have this setting on, then in main window you will see a capture button at the bottom. This allows you to save the recorded signal of the last 1,2,3,... seconds in a wav-file. I introduced it to allow users to easily record a signal and then provide the sound file for analysis (e.g. the tuner is not working well for them). But up to now, this was only used by me ...

I tested it and found what it does, but still a but unclear to me, for what is it used? What you do with these records?

> Please let me know, if this helps ...

You answers already helped a lot, altough many new questions, I'm getting to...

Very cool that this tuner contain such audio analyse tools. They have already helped me with a couple of tasks and I'm so interested in audio analysing, but still a lot to learn as you can see :)

**Comment by thetwom** [comment]
Ok, another try:

*Tolerance*
This is given in cents. One cent is a hundredth of a half tone in equal temperament. Hz cannot be used here, since the difference in Hz between two notes is much smaller for low notes than for high notes. 

*Pitch*
Yes, jump to next pitch after 5 similar results, means, that we only change target note, when e.g. for 5 times 0.05s evaluations give a similar result. A pitch result is the "most probable pitch" (but not strictly in a mathematical sense). This is not necessarily the strongest frequency. As said, a note can contain higher harmonics, so A4 might not only have 440Hz, but also 880Hz, 1220Hz, .... And the 440Hz must not necessarily the strongest. But the combination of all harmonics should be the strongest. 

*Background noise*
It is the other way round. If you set it to 0%, the tuner will never try to find a tone, since it will allow no noise at all. And if you set it to 100%, the tuner will always try to find a tone. Assume, you are not playing a note. Of course, your mobile will still get an audio signal since there is always some noise. The tuner can of course try to find tones on that, but it would just jump around without meaningful tones. So we need a mechanism to tell "quietness" from "playing a note" apart. At the moment, the tuner is doing that by comparing the noise level to the loudness of meaningful signals. 

*Spectrum*
Yes the spectrum monitor shows "real-time" the current frequency content of a signal. No time-domain signal. The highest frequency can be the pitch, but as said before, it is the combination of harmonics. So e.g. for A4, you could also have no 440Hz peak at all but instead just 880Hz, 1220Hz, .... The difficulty is then of course not to detect A5, but this would e.g. not contain 1220Hz because it is not a multiple of 880Hz. 

*Autocorrelation*
Horizontal axis is strictly speaking the time shift of the signal. So if the signal is shifted by a certain amount and compared to the original signal, how similar they are. But a time shift is related to a frequency by its inverse. So instead of labeling the ticks as time shifts of 0.1s, 0.2s, ... we show 1/0.1s, 1/0.2s, ... which is the frequency. The vertical axis is not the "strength of Hz", but the similarity of the shifted signal to the original signal. The peaks of the "snake" will decrease towards the right since we compare signals, which are not going on for ever, but e.g. only 0.1s. So if you shift a 0.1s by e.g. 0.05s, half of the signal is not overlapping anymore, reducing the similarity. To get a better understanding of this, you should try compute such values by yourselves. You don't need instrument signals for that. You can start with simple sine waves. 

*Windowing*
You can e.g. ask wikipedia about windowing functions, seeing their shapes. In the tuner you are right, that the results is absolutely not sensitive to the windows. As said, possibly windowing is not needed for a tuner. 

*Capture*
The tuner analyses sound real-time. This makes it hard to investigate, when it works, and when not. If you have a stored signals, you can analyse them carefully, look closely at autocorrelation and spectra and understand, why the tuner correctly finds a pitch or not. E.g. if you are trying to tune some new instrument, but the tuner does not give good results, you could send me a sound file, and I could try to find out, why it fails. Without such an example file there it becomes very hard to understand, why the tuner fails .

*Learning about signal processing*
I really cannot go into the deep details here. You have a lot of ideas here, where you can start deeper research. Start with Fourier transform and correlation. Try to analyse very simple signals by yourselves, like sine waves. When this works, you can try to analyse more realistic signals.

**Comment by SadeeshaJayaweera** [comment]
It seems you have several questions about technical aspects related to audio processing and analysis. Let me provide some explanations for each of your questions:

1. **Tolerance Setting:**
   - In the context of audio processing and pitch analysis, tolerance typically refers to the acceptable variation or deviation from a reference pitch. It's a threshold value that determines how closely a detected pitch must match the reference pitch for it to be considered a valid result.
   - For example, if you have a reference pitch of 440 Hz (A4), and you set a tolerance of 10 Hz, any detected pitch within the range of 430 Hz to 450 Hz would be considered a match.

2. **Pitch Result:**
   - Pitch result generally refers to the output or outcome of a pitch detection or analysis process. It represents the estimated pitch value of an audio signal.
   - Different algorithms and methods can be used to calculate pitch, such as autocorrelation, spectral analysis, or others. The "pitch result" is the pitch value obtained through these methods.

3. **Background Noise:**
   - Background noise in the context of audio analysis typically refers to any unwanted or ambient sound present in an audio recording that is not part of the primary signal of interest.
   - Background noise can include things like room noise, electrical interference, or any sound not produced by the intended source. It's important to remove or filter out background noise to improve the accuracy of audio analysis.

4. **Spectrum vs. Autocorrelation:**
   - **Spectrum** refers to the distribution of signal energy across different frequency components. It provides information about the frequency content of an audio signal. Methods like the Fast Fourier Transform (FFT) are used to calculate the spectrum.
   - **Autocorrelation** is a technique used to find repeating patterns or periodicity in a signal. It involves comparing a signal with delayed versions of itself to identify repeating features.
   - The key difference is that spectrum analysis provides information about frequency components, while autocorrelation is more focused on identifying periodic patterns or pitch in a signal.

5. **Windowing Function:**
   - A windowing function is used in signal processing to taper or shape a segment of a signal before applying operations like the Fourier Transform. It's used to reduce spectral leakage and other artifacts.
   - Window functions, such as the Hamming window or Blackman window, gradually reduce the amplitude of the signal towards its ends, reducing the impact of sharp transitions.
   - Windowing functions help in analyzing short segments of a signal while minimizing artifacts caused by abrupt signal changes at the edges of the segment.

6. **Capture in Scientific Mode:**
   - "Capture in Scientific Mode" likely refers to a feature in an audio analysis tool or software. It's used to capture audio data for scientific analysis and experimentation.
   - Scientific mode often provides more detailed and customizable options for audio capture, allowing researchers and scientists to collect data for various types of analysis, including pitch analysis, spectral analysis, and more.

Please note that the precise definitions and functionalities of these terms may vary depending on the specific audio analysis software or tool you are using, so it's a good practice to refer to the documentation or user manual of the particular application for detailed information.

**Comment by Glandos** [comment]
This seems to be a nice start for a FAQ. Even within the application itself (but this would require translation. Err wait, I'm a translator, arg!)
